### ISSUE TITLE
Slow performance with shadow dom

### ISSUE THREAD

**Issue by paladox** [codemirror/CodeMirror#6758]
I'm not exactly sure if this is an bug within CodeMirror or we're doing something wrong. But basically when we insert large amount of content (e.g. https://raw.githubusercontent.com/GerritCodeReview/gerrit/master/Documentation/rest-api-changes.txt for example), scrolling and clicking is slow. I've managed to reproduce in non-shadow-dom too if using height:auto on .CodeMirror. You can see here https://codepen.io/edgenard/pen/mEWeNX (you can try and insert the content first without the height:auto then with it. Notice the performance difference?). There is this codepen (shadow dom) https://codepen.io/ilyagladko/details/KKdXamm which once you insert the content you can see the performance issue.

We use shadow dom for our element codemirror-editor (https://github.com/GerritCodeReview/plugins_codemirror-editor/blob/master/gr-editor/gr-editor.js & https://github.com/GerritCodeReview/plugins_codemirror-editor/blob/master/gr-editor/codemirror-element.js).

We use the height:auto on our element but removing it does not fix the issue though.

**Comment by paladox** [comment]
Found why, it's because of `viewportMargin: Infinity` but then why does height: auto cause the performance issue too?

**Comment by paladox** [comment]
Oh I guess it's just doing viewportMargin infinity if the height is auto? That seems like a bug since it should use the default value or the one you set.

**Comment by marijnh** [comment]
> Oh I guess it's just doing viewportMargin infinity if the height is auto?

It's not really, in that that option isn't being overridden, it's just that if the editor has the height of the content, viewporting doesn't work (since it relies on the editor's scrollbox).

[CodeMirror 6](https://codemirror.net/6/) has a more general viewporting implementation that'll also use the window's scrollbox, but this is unlikely to ever be added to version 5.x.

**Comment by paladox** [comment]
I've found a hacky solution for this issue: https://gerrit-review.googlesource.com/c/plugins/codemirror-editor/+/313422.

Basically uses the search add-on so users can still search the entire document... but we also set the size (height) using setSize. It's hacky as if you have for example the developer tools open the size is much smaller (and doesn't resize when you close it). But it does the job I want
